### PR TITLE
Consolidate make_rpm.bzl and experimental/make_rpm.bzl

### LINKED ISSUE #173: Consolidate experimental and existing make_rpm scripts
As @aiuto [mentioned](https://github.com/bazelbuild/rules_pkg/pull/129#issuecomment-597897774) in #129, there were some modifications made to the `make_rpm` script that were committed.

It would be nice if both `pkg_rpm`'s were able to use the same RPM generation script under the hood to reduce the overall maintenance burden, until if/when the way packaging is done under the hood diverges significantly.

### PR DESCRIPTION
**Pull request by nacl**
This change consolidates all of new features/fixes from
pkg/experimental/make_rpm.bzl into pkg/make_rpm.bzl.

Code used specifically by experimental/rpm.bzl is marked as such, and is grouped
as to be generally obvious.

Features now supported by non-experimental `pkg_rpm`:
- Support for older versions of `rpm(8)` (creation of `RPMS` output directory)

Features supported by non-experimental `pkg_rpm` not supported by experimental
`pkg_rpm`:
- Support for `SOURCE_DATE_EPOCH`

Fixes #173.

### CODE REVIEW

File: pkg/make_rpm.py [hunk 1]
```diff
@@ -199,36 +228,106 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+
+    # Further, the use of disabling yapf and friends is to allow argument names
+    # to be associated with their values neatly.
     args = [
-        self.rpmbuild_path,
-        '--define',
-        '_topdir %s' % dirname,
-        '--define',
-        '_tmppath %s/TMP' % dirname,
+      self.rpmbuild_path,  # noqa: E121
```
Comment (aiuto) [inline pkg/make_rpm.py:295]: Why would you have to re-indent everything. The args are already 4 space indent, like at line 301.

File: pkg/make_rpm.py [hunk 2]
```diff
@@ -298,20 +408,55 @@ def main(argv):
...
+                         # Additional scriptlet path arguments set when being
+                         # used by experimental/rpm.bzl
+                         pre_scriptlet_path=options.pre_scriptlet,
+                         post_scriptlet_path=options.post_scriptlet,
+                         preun_scriptlet_path=options.preun_scriptlet,
+                         postun_scriptlet_path=options.postun_scriptlet)
     builder.AddFiles(options.files)
-    return builder.Build(options.spec_file, options.out_file)
+    return builder.Build(options.spec_file, options.out_file,
+                         # Additional arguments set when being used by
```
Comment (aiuto) [inline pkg/make_rpm.py:448]: I don't see a need to say experimental here. If we put them in the main make_rpm, they are part of the interface. OTOH, they are a secret part, so I don't feel bad about breaking people who depend on internals. 
But... we intend these to be permanent features, so why not just have them as args, with no mention of experimental?

Comment (nacl) [reply pkg/make_rpm.py:448]: I buy it.  I'll clean up the comments.

File: pkg/make_rpm.py (hunk 1 above)
Comment (nacl) [reply pkg/make_rpm.py:295]: The thought was to make it consistent with the rest of the code.  It's easy enough to revert it, though.

The whole file should probably be changed to use a four-space indent, TBH.

File: pkg/make_rpm.py [hunk 3]
```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+      tpl_replacements = {}
+      if self.version:
+        tpl_replacements['VERSION_FROM_FILE'] = self.version
+      if self.release:
+        tpl_replacements['RELEASE_FROM_FILE'] = self.release
+      CopyAndRewrite(os.path.join(original_dir, preamble_file),
+                     self.preamble_file,
+                     template_replacements=tpl_replacements)
+
+    self.description_file = None
```
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py:269]: > This feels slightly weird, for a few reasons.
> 
> We have an RpmBuilder class, but this group of a few class members (*_file) are not initialized in the **init** method. Just to keep that clean, these options could be passed to the class initializer and pre-set there.
> `self.description_file = os.path.basename(description_file)` if description_file or None
> 
> So when we reach this point it might look like
> 
> ```python
> if self.description_file:
>     shutil.copy(self.description_file, os.getcwd())
> ```
> 
> But this is obviously different than what you have, but I can't quite understand that.

At the time, I believe was trying to be consistent with the surrounding code (see `self.spec_file`), which has some values that are not defined in the constructor.

Now that I think on this a little, all attributes in the class should probably be initialized to some default value even if they are to actually be set later, as is required by many more "static" programming languages.  This could avoid a bunch of spaghetti that would otherwise show up when you define everything.

> Maybe this is just me forgetting rpmbuild behavior, but maybe the code needs some annotation to explain why certain things are done.
>
>     * It seems we move some files to the build dir even though we include them, but we have the full paths. Can we not %include with a full path.  And maybe we are doing that.

I honestly don't know whether or not `%include` can accept absolute paths.  TBH, I never tried.

>       I see  shutil.copy(os.path.join(original_dir, install_script_file), os.getcwd())
>       which feels odd.  It seems that in the rule impl we had the full path to the install_script_file, then we basename that somewhere, then add in the path of the dir we started in. It's hard to follow.  Maybe it just needs some comments in the code about what original_dir means and the significance of were we chdir 

Comments are easy enough to do.  I'll write these in the immediate term.

However, you are very much correct in that this is a confusing mess.  There are currently two ways to inject data into the RPM spec template:

1) A shell-script-style ${VARIABLE} template, which is expanded into the RPM spec file directly
2) A rpm-style macro statement, e.g. `%include %build_rpm_install`, which are expanded inline by`rpmbuild` when it is called.

I get the feeling that this whole process would be greatly simplified if this process was broken down into two or three distinct phases:

1) The template spec file is expanded to a more fully-qualified spec file that can be invoked by `rpmbuild` with minimal additional magic.
2) The rpmbuild directory structure is created, and, if necessary (see below), the already-built artifacts are put in the right place.
3) `rpmbuild` is invoked with the necessary arguments to build the RPM (This would be `make_rpm.py`)

This could help us dramatically reduce the complexity of `make_rpm.py`.

We might be able to somewhat reduce the number of file copies this process entails by having the `%install` phase simply copy (or hardlink) the files from the bazel build tree to where they are supposed to be relative to RPM builds the package.  Haven't yet tried this out either.

[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py:269]: Tracked in #209.

File: pkg/make_rpm.py [hunk 4]
```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+    if description_file:
+      shutil.copy(os.path.join(original_dir, description_file), os.getcwd())
```
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py:271]: Yeah, this is fair, but I think the mechanism for global template expansion needs to be thought out a little more.  For example, users may also want to embed version numbers into `pkg_filegroup` paths too.

The whole "how do you do the template" question needs to be solved to make the interface "nicer".

This is out-of-scope for this change, though.  Tracked in #210.

File: pkg/make_rpm.py [hunk 5]
```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+                   template_replacements=tpl_replacements)
+
+    # preamble replacement, %description, %install, %files -f substitution
+    self.preamble_file = None
+    if preamble_file:
+      # Copy in the various other files needed to build the RPM
+      self.preamble_file = os.path.basename(preamble_file)
+      tpl_replacements = {}
+      if self.version:
+        tpl_replacements['VERSION_FROM_FILE'] = self.version
```
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py:262]: The reasoning for the name was that the Starlark rule will have already done the substitution if it the version is instead provided as an attribute to `pkg_rpm`.

While you are correct that the the template really shouldn't care where the value comes from, at this time, the preamble template file is not controllable via a public interface, so the existing name is harmless.

File: pkg/make_rpm.py [hunk 6]
```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+        '--define', '_tmppath %s/TMP' % dirname,
         '--bb',
         '--buildroot=%s' % buildroot,
-        self.spec_file,
-    ]
+    ]  # yapf: disable
+
+    # Macro-based RPM parameter substitution, if necessary inputs provided.
+    if self.preamble_file:
+      args += ['--define', 'build_rpm_options %s' % self.preamble_file]
```
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py:310]: The style is consistent when you look at it from the perspective of single vs. multiple simultaneous additions to the list; the example you point to adds multiple.  I also use append on line 297 (`args.append('-vv')` for debugging).

What's the ask here?  Should everything use `append()`, or `+= [items...]`, or is it fine as it is?

File: pkg/make_rpm.py [hunk 7]
```diff
@@ -199,36 +230,103 @@ def SetupWorkdir(self, spec_file, original_dir):
...
+    if self.install_script_file:
+      args += ['--define', 'build_rpm_install %s' % self.install_script_file]
+    if self.file_list_file:
+      # %files -f is taken relative to the package root
+      args += ['--define', 'build_rpm_files %s' % os.path.basename(self.file_list_file)]
+
+    args.append(self.spec_file)
+
+    if self.debug:
+      print('Running rpmbuild as: {}'.format(' '.join(["'" + a + "'" for a in args])))
```
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py]: Addressed in 9f906b39161e5092ea3ebf8143c64344b2618715.

File: pkg/make_rpm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nacl) [reply pkg/make_rpm.py]: I elected to move the scriptlets out of the constructor and put them into `SetupWorkdir` instead.  All the values are initialized to `None` in the constructor, though.

See 9f906b39161e5092ea3ebf8143c64344b2618715.

File: pkg/make_rpm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aiuto) [reply pkg/make_rpm.py:352]: We're both right!  I say it is one arg, and you say it is multiple words.  I guess neither is better than the other here.
